### PR TITLE
Simplify reverse tuple iteration

### PR DESCRIPTION
**Pull request by vittorioromeo**
Should help a bit with #1160. Would also be nice to see how it interacts with #1055

### CODE REVIEW

**Comment by fnc12** [comment]
@vittorioromeo thank you for this PR. Please resolve the conflicts to be able to run CI tests

**Comment by vittorioromeo** [comment]
> @vittorioromeo thank you for this PR. Please resolve the conflicts to be able to run CI tests

Should be fine now :)

**Comment by fnc12** [comment]
@vittorioromeo did you test compilation time with your branch? I am actually not sure whether reverse tuple iteration is used anywhere so I am not sure whether your fix will make much effort

**Comment by vittorioromeo** [comment]
> @vittorioromeo did you test compilation time with your branch? I am actually not sure whether reverse tuple iteration is used anywhere so I am not sure whether your fix will make much effort

I saw a very small improvement of average 24ms per instantiation of `iterate_tuple` from 28ms, but hard to tell if it's impactful in any way. There are a few calls to `iterate_tuple<true>` in the codebase. Regardless, I think that the PR is an improvement overall because of the removal of unnecessary code.

File: dev/functional/index_sequence_util.h
Comment (trueqbit) [inline dev/functional/index_sequence_util.h:1]: Could you please also remove
```c++
#ifdef SQLITE_ORM_RELAXED_CONSTEXPR_SUPPORTED
#include <array>
#endif
```
from the top of the header?

File: dev/tuple_helper/tuple_iteration.h [hunk 1]
```diff
@@ -33,7 +33,8 @@ namespace sqlite_orm {
         template<bool reversed = false, class Tpl, size_t... Idx, class L>
         void iterate_tuple(const Tpl& tpl, std::index_sequence<Idx...>, L&& lambda) {
             if constexpr(reversed) {
-                iterate_tuple(tpl, reverse_index_sequence(std::index_sequence<Idx...>{}), std::forward<L>(lambda));
+                int sink;
```
Comment (trueqbit) [inline dev/tuple_helper/tuple_iteration.h:36]: Please put a comment above similar to:
`// nifty fold expression trick: make use of guaranteed right-to-left evaluation order when folding over operator=`.

Even seasoned C++ programmers will find it helpful :) - Even I had to try it out and find out about it.

**Comment by trueqbit** [comment]
This is awesome C++! Thanks @vittorioromeo for thinking along.

I doubt you will see any overall improvement in compilation times.
But it simplifies the code, and that was what I was originally looking for when I updated the tuple iteration but couldn't find it!

By the way, there is good reading about this fold expression trickery:
* https://www.foonathan.net/2020/05/fold-tricks
* https://quuxplusone.github.io/blog/2020/05/07/assignment-operator-fold-expression/

**Review by trueqbit — CHANGES_REQUESTED** [review]

File: dev/functional/index_sequence_util.h
Comment (trueqbit) [inline dev/functional/index_sequence_util.h:1]: BTW, we might implement `first_index_sequence_value()` as follows. What do you think?

```c++
template<size_t... Idx>
SQLITE_ORM_CONSTEVAL size_t first_index_sequence_value(std::index_sequence<Idx...>) {
    static_assert(sizeof...(Idx) > 0);
    size_t result;
    ((result = Idx, true) || ...);
    return result;
}
```

Comment (vittorioromeo) [reply dev/functional/index_sequence_util.h:1]: Compared to the current implementation, I'm not sure it's going to be faster. I would recommend a benchmark on https://www.build-bench.com/

**Review by trueqbit — APPROVED** [review]

**Review by fnc12 — APPROVED** [review]

**Comment by fnc12** [comment]
@vittorioromeo thanks!

**Pull request merged**